testing/unit/munit-open-close.cpp

```diff
@@ -0,0 +1,158 @@
+
+using namespace opae::testing;
+
+class openclose_c_p
+    : public ::testing::TestWithParam<std::string> {
+ protected:
+  openclose_c_p() : tmpsysfs_("mocksys-XXXXXX"){}
+
+  virtual void SetUp() override {
+    ASSERT_TRUE(test_platform::exists(GetParam()));
+    platform_ = test_platform::get(GetParam());
+    system_ = test_system::instance();
+    system_->initialize();
+    tmpsysfs_ = system_->prepare_syfs(platform_);
+
+    ASSERT_EQ(fpgaGetProperties(nullptr, &filter_), FPGA_OK);
+    ASSERT_EQ(fpgaPropertiesSetObjectType(filter_, FPGA_DEVICE), FPGA_OK);
+    ASSERT_EQ(fpgaEnumerate(&filter_, 1, tokens_.data(), tokens_.size(),
+                            &num_matches_),
+              FPGA_OK);
+  }
+
+  virtual void TearDown() override {
+    EXPECT_EQ(fpgaDestroyProperties(&filter_), FPGA_OK);
+    if (!tmpsysfs_.empty() && tmpsysfs_.size() > 1) {
+      std::string cmd = "rm -rf " + tmpsysfs_;
+      std::system(cmd.c_str());
+    }
+    system_->finalize();
+  }
+
+  std::string tmpsysfs_;
+  fpga_properties filter_;
+  std::array<fpga_token, 2> tokens_;
+  uint32_t num_matches_;
+  test_platform platform_;
+  test_system *system_;
+};
+
+
+
+
+/**
+ * @test       open_01
+ *
+ * @brief      When the fpga_handle * parameter to fpgaOpen is NULL, the
+ *             function returns FPGA_INVALID_PARAM.
+ */
+TEST_P(openclose_c_p, open_01) {
+  fpga_token tok = NULL;
+  fpga_result res;
+  res = fpgaOpen(tok, NULL, 0);
+  ASSERT_EQ(FPGA_INVALID_PARAM, res);
+}
+
+/**
+ * @test       open_02
+ *
+ * @brief      When the fpga_handle * parameter to fpgaOpen is NULL, the
+ *             function returns FPGA_INVALID_PARAM.
+ */
+TEST_P(openclose_c_p, open_02) {
+  fpga_token tok;
+  fpga_result res;
+  res = fpgaOpen(tok, NULL, 0);
+  ASSERT_EQ(FPGA_INVALID_PARAM, res);
+}
+
+/**
+ * @test       open_03
+ *
+ * @brief      When the fpga_token parameter to fpgaOpen is NULL, the
+ *             function returns FPGA_INVALID_PARAM.
+ */
+
+TEST_P(openclose_c_p, open_03) {
+  fpga_handle handle_;
+  ASSERT_EQ(FPGA_INVALID_PARAM, fpgaOpen(NULL, &handle_, 0));
+}
+
+
+/**
+ * @test       open_04
+ *
+ * @brief      When the flags parameter to fpgaOpen is invalid, the
+ *             function returns FPGA_INVALID_PARAM.
+ *
+ */
+TEST_P(openclose_c_p, open_04) {
+  fpga_handle handle_;
+  ASSERT_EQ(FPGA_INVALID_PARAM, fpgaOpen(tokens_[0], &handle_, 42));
+}
+
+/**
+ * @test       open_05
+ *
+ * @brief      When the flags parameter to fpgaOpen is invalid, the
+ *             function returns FPGA_INVALID_PARAM and FPGA_NO_DRIVER.
+ *
+ */
+TEST_P(openclose_c_p, open_05) {
+  fpga_handle handle_;
+  fpga_result res;
+  struct _fpga_token* _token = (struct _fpga_token*)tokens_[0];
+
+  res = fpgaOpen(tokens_[0], &handle_, 42);
+  ASSERT_EQ(FPGA_INVALID_PARAM, res);
+
+  _token->magic = FPGA_TOKEN_MAGIC;
+  res = fpgaOpen(tokens_[0], &handle_, FPGA_OPEN_SHARED);
+  ASSERT_EQ(FPGA_OK, res);
+
+  strcpy(_token->devpath,"/dev/intel-fpga-fme.01");
```

Comment:
Did you intend to use intel-fpga-fme.01?  The leading 0 makes the path invalid.

Reply:
I did that intentionally to cover one of the cases.

Reply:
Ok.. We still need to move away from peeking inside the internal structures.  The handles and tokens we get back from the plugin api have an extra layer on top of the _fpga_token and _fpga_handle structures.